scvi/external/stereoscope/_module.py

```diff
@@ -0,0 +1,240 @@
+        self.W = torch.nn.Parameter(torch.randn(self.n_genes, self.n_labels)) # n_genes, n_cell types
+
+    def get_params(self) -> torch.Tensor:
+        """
+        Returns the parameters for feeding into the spatial data
+        -------
+        type
+            list of tensor
+        """
+        return self.W.detach().cpu().numpy(), self.px_o.detach().cpu().numpy()
+
+    def _get_inference_input(self, tensors):
+        # we perform MAP here, so there is nothing to infer
+        return {}
+
+    def _get_generative_input(self, tensors, inference_outputs):
+        x = tensors[_CONSTANTS.X_KEY]
+        y = tensors[_CONSTANTS.LABELS_KEY]
+
+        input_dict = dict(
+            x=x, y=y
+        )
+        return input_dict
+
+
+    @auto_move_data
+    def inference(self):
+        return {}
+
+    @auto_move_data
+    def generative(
+        self, x, y
+    ):
+        """Simply build the negative binomial parameters for every cell in the minibatch."""
+        px_scale = torch.nn.functional.softplus(self.W)[:, y.long()[:, 0]].T # cells per gene
+        library = torch.sum(x, dim=1, keepdim=True)
+        px_rate = library * px_scale
+
+        return dict(
+            px_scale=px_scale,
+            px_o=self.px_o,
+            px_rate=px_rate,
+            library=library,
+        )
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        kl_weight: float = 1.0,
+    ):
+        x = tensors[_CONSTANTS.X_KEY]
+        px_rate = generative_outputs["px_rate"]
+        px_o = generative_outputs["px_o"]
+
+        reconst_loss = -NegativeBinomial(px_rate, logits=px_o).log_prob(x).sum(-1)
+        loss = torch.mean(reconst_loss)
+
+        return SCVILoss(loss, reconst_loss, torch.zeros((1,)), 0.0)
+
+    @torch.no_grad()
+    def sample(
+        self,
+        tensors,
+        n_samples=1,
+        library_size=1,
+    ):
+    # not a generative model
+        return None
+
+class SpatialDeconv(AbstractVAE):
+    """
+    Model of single-cell RNA-sequencing data for deconvolution of spatial transriptomics:
+    Reimplementation of the STModel module of Stereoscope : https://github.com/almaan/stereoscope/blob/master/stsc/models.py.
+
+    Parameters
+    ----------
+    n_spots
+        Number of input spots
+    params 
+        Tuple of ndarray of shapes [(n_genes, n_labels), (n_genes)] containing the dictionnary and log dispersion parameters
+    """
+    def __init__(
+        self,
+        n_spots: int,
+        params: Tuple[np.ndarray],
+        prior_weight:Literal["n_obs", "minibatch"] = "n_obs",
+    ):
+        super().__init__()
+        self.W = torch.nn.Parameter(torch.tensor(params[0]), requires_grad=False)
+        self.px_o = torch.nn.Parameter(torch.tensor(params[1]), requires_grad=False)
+        self.n_spots = n_spots
+        self.n_genes, self.n_labels = self.W.shape
+        self.prior_weight = prior_weight
+
+        #####
```

Comment:
this math could actually go in the class docstring, if you look at the numpydoc specification, there's a notes section and you can do block math latex.

Reply:
this applies to all the math if you wish.